Add Python linting via hound-python service.

jmcarp opened the pull request:
Add support for Python linting via https://github.com/jmcarp/hound-python. Replaces #754.

teoljungberg commented on app/models/style_checker.rb:
```diff
@@ -44,6 +44,8 @@ def style_guide_class(filename)
       StyleGuide::Haml
     when /.+\.scss\z/
       StyleGuide::Scss
+    when /.+\.py\z/
```
Are we missing a test for this?

teoljungberg replied:
We don't, nevermind

teoljungberg commented:
This PR looks good, though I have some ideas for the linter itself.

One of the goals we had for this new achitechture was to write the linters in their own language. Like done in https://github.com/thoughtbot/hound-go

Could we do that for your python linter aswell? I'm not a python developer myself, so maybe we can get someone else from our side to review the code? @mike-burns have talked about implementing the same linter and should be able to help you.

Any other ideas @thoughtbot/hound ?

jmcarp commented:
Writing the go linter in go is straightforward because there's already a resque client written in go. I looked around but haven't found a python resque worker implementation--probably because most python devs use celery or rq. I could write the linter in python, but would have to implement the interface to resque. Which wouldn't be a whole lot of work, but is definitely more complicated than shelling out to `flake8`, like I'm doing now.

Is that more like what you had in mind @teoljungberg?

jmcarp commented:
Here's what a python implementation of the python linter might look like. I'd be happy to flesh this out it seems like the right direction for the task.
```python
# resque.py

import time
import collections

def _get_key(queue):
	return 'resque:queue:{0}'.format(queue)

class Resque(object):

	def __init__(self, conn, interval=5):
		self.conn = conn
		self.interval = interval
		self._registry = collections.defaultdict(dict)

	def work(self):
		while True:
			for queue in self._registry:
				key = _get_key(queue)
				raw = self.conn.lpop(key)
				if raw:
					body = json.loads(task)
					task = self._registry[body['class']]
					try:
						task(*body['args'])
					except Exception as error:
						pass
			time.sleep(self.interval)

	def enqueue(self, queue, klass, *args):
		key = _get_key(queue)
		value = {
			'class': klass,
			'args': args,
		}
		self.conn.rpush(key, json.dumps(value))

	def task(self, queue, klass):
		def wrapper(func):
			self._registry[queue][klass] = func
			return func
		return wrapper

# tasks.py

import pep8
import flake8.main as flake8

import redis

import resque

conn = redis.StrictRedis()
resque = resque.Resque(conn)

@resque.task('python_review', 'PythonReviewJob')
def python_review(filename, commit_sha, pull_request_number, patch, content, config):
	violations = [
		{
			'line': each[0],
			'message': each[3]
		}
		for each in check_code(content)
	]
	resque.enqueue(
		'high',
		'CompletedFileReviewJob',
		filename,
		commit_sha,
		pull_request_number,
		patch,
		violations,
	)

def check_code(code, **kwargs):
    flake8_style = flake8.get_style_guide(**kwargs)
    flake8_style.options.report = QuietReport(flake8_style.options)
    return flake8_style.input_file(None, lines=code.splitlines(True))

class QuietReport(pep8.StandardReport):
	def get_file_results(self):
		return sorted(self._deferred_print)

if __name__ == '__main__':
	resque.work()
```

jmcarp commented:
Actually, I spoke too soon. There *is* a python clone of resque called pyres. It doesn't appear to be actively maintained, and there are a few differences from the original resque implementation, but it looks close enough that it might work here. I'll take a crack at rewriting the python linter with pyres and see how it goes.

mike-burns commented:
@teoljungberg this PR is for making use of an existing external service, like what we do with SCSS. I have not been following Hound closely, but isn't this what we want?

teoljungberg commented:
@mike-burns it is, but we would prefer having the linters written in their own language. So a JS linter would be written in JS, a Go linter would be written in Golang, and a Python linter would be written in Python.

If it's impossible to write the linter in it's own language, like for SCSS(if someone writes a Resque implementation for SCSS, I'm impressed), we would fallback to Ruby.

salbertson commented:
@jmcarp what do you think about disabling Python by default until we know it's stable in production?

I did this recently with SCSS review by disabling all rules in the config. Is there something similar we could do in https://github.com/jmcarp/hound-python?

salbertson commented:
@jmcarp I'm totally fine with the service being written mostly in Ruby.

jmcarp commented:
Now that I've thought about it a bit, it makes sense to write linters in their own languages. Plus, I'm better with Python than Ruby. I put together a Python linter in Python at https://github.com/jmcarp/hound-python (I still have the old code if you wind up preferring Ruby). I'll document the linter this week, but for now, it's working in development, and tests are passing.

@salbertson I'll add a default config with all rules disabled.

jmcarp commented:
The Python linter has been updated with basic documentation, and now ignores all files by default. To enable, users would pass a flake8 ini-style configuration file like this:
```
[flake8]
exclude=
```
If anybody has time to look over the Python linter, I'd be happy to revise as necessary.

mike-burns commented on .env:
```diff
@@ -16,3 +16,8 @@ SECRET_KEY_BASE=some_really_really_long_super_secret_key_base
+GITHUB_CLIENT_SECRET=d163570955e4c95012b796911d7abb4729b32dc7
+HOST=jmc-hound.ngrok.com
+HOUND_GITHUB_TOKEN=f0c80f95dd2d4770b8a54e8d8ab3ab5ee3f0b96f
+HOUND_GITHUB_USERNAME=jmcarp
```
Can you change these values to one that look more obviously like dummy data?

jmcarp replied:
Sorry, reverted.

On Wed, Jul 22, 2015 at 5:23 AM, Mike Burns <notifications@github.com>
wrote:

> In .env
> <https://github.com/thoughtbot/hound/pull/853#discussion_r35195048>:
>
> > @@ -16,3 +16,8 @@ SECRET_KEY_BASE=some_really_really_long_super_secret_key_base
> >  SPLIT_ADMIN_PASSWORD=split_password
> >  STRIPE_API_KEY=sk_test_123
> >  STRIPE_PUBLISHABLE_KEY=pk_test_123
> > +GITHUB_CLIENT_ID=43d612c5c4251c68574e
> > +GITHUB_CLIENT_SECRET=d163570955e4c95012b796911d7abb4729b32dc7
> > +HOST=jmc-hound.ngrok.com
> > +HOUND_GITHUB_TOKEN=f0c80f95dd2d4770b8a54e8d8ab3ab5ee3f0b96f
> > +HOUND_GITHUB_USERNAME=jmcarp
>
> Can you change these values to one that look more obviously like dummy
> data?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/thoughtbot/hound/pull/853/files#r35195048>.
>

mike-burns commented:
I'm glad this is a light wrapper around flake8. I've no concrete feedback on the Python code. CC @thoughtbot/python .

salbertson commented:
> but we would prefer having the linters written in their own language

@teoljungberg I don't think I agree. Why would we prefer having the linters written in their own language?

I would rather have the style services written in Ruby, if that's a realistic option, since the services will most likely be maintained by thoughtbot and Ruby is the language most of us are more familiar with.

jmcarp commented:
Having written hound-python in both Ruby and Python, I have a slight preference for the native approach. Most linters will be implemented in the language they're meant to check, so writing all the services in Ruby means managing cross-language dependencies and talking to linters via subprocess. Which means writing potentially brittle code to prepare arguments for subprocess, parse output strings, handle errors, etc.

Regarding maintenance, linters should be trivial--the Python implementation of hound-python is something like 50 lines, plus tests and some utility code. So writing native linters wouldn't saddle you with maintaining large or complicated non-Ruby codebases. Plus, I would guess that you'd get more help from the community on maintenance with native linters--lots of folks in the Python community would help fix bugs in a linter written in Python, but not so many would be able to help with a Ruby implementation.

All that said, feel free to use the Ruby implementation of the Python linter if you prefer. It's at https://github.com/jmcarp/hound-python-ruby.

salbertson commented:
Did you run into many issues with brittle code while writing the Ruby version?

jmcarp commented:
Some. The code is so simple that I'm not too concerned, but here's a few examples of what I mean:
```ruby
def lint(content, config = {})
  opts = config.to_a.map { |key, value| "--#{key} #{value}" }.join(" ")
  report = `echo "#{content}" | flake8 - #{opts}`
  report.split("\n").map do |line|
    parse_flake8_line(line)
  end
end

def parse_flake8_line(line)
  parts = line.split(":")
  {
    line: parts[1].to_i,
    message: parts.drop(3).join(":").strip
  }
end
```
This passes all the options from the config file to the `flake8` executable, which has to be on the path. If the config file has invalid options, the subprocess call will fail. Then we have to parse the output of `flake8` as a string and munge it into line number and message. If for some reason the delimiting characters or position of different components change in `flake8`, we'll munge the errors incorrectly. Neither if these issues comes up with a native implementation. Anyway, now that we're talking about it, we would need to add some error handling to the Ruby implementation if we go in that direction.

croaky commented:
> Why would we prefer having the linters written in their own language?

I'm in favor of preferring native, if only for the community reason:

> get more help from the community on maintenance with native linters--lots of folks in the Python community would help fix bugs in a linter written in Python, but not so many would be able to help with a Ruby implementation.

Our job is to keep the linter API simple and well-defined so running these on Heroku and accessing the Redis queue is straightforward to operationalize, but make each linter as comfortable as possible for its language community.

salbertson commented:
@jmcarp ok, sounds like Python based service is a better choice. Would you mind using double quotes and then squashing your commits. This looks good to me. :+1:

gylaz commented on app/models/style_guide/python.rb:
```diff
@@ -0,0 +1,26 @@
+      Resque.push(
+        'python_review',
+        'class' => 'review.PythonReviewJob',
+        'args' => [
```
Mind changing this to ruby 1.9 style hashes?

gylaz commented on spec/models/style_checker_spec.rb:
```diff
@@ -167,6 +167,17 @@
 
+  context "for a Python file" do
+    it "does not immediately return violations" do
+      commit_file = stub_commit_file("test.py", "* { color: red; }")
```
How about updating the content string to be something Python?

gylaz commented on spec/models/style_checker_spec.rb:
```diff
@@ -167,6 +167,17 @@
+
+      expect(violation_messages).to be_empty
+    end
+  end
```
I wonder what the value of this test (and the one it was based on) is? Maybe a better test would be to check that the associated `file_review` is pending. I'm also okay if we handle this in a different PR.

jmcarp commented:
Commits squashed, merge conflicts resolved.

salbertson commented:
Merged! Thanks @jmcarp for pushing this along!

salbertson commented:
@jmcarp Would you mind transferring the https://github.com/jmcarp/hound-python repo to the thoughtbot org and I'll make you a contributor?

salbertson commented:
@jmcarp, what do you think about transferring the service repo?

thorncp commented:
@jmcarp Any updates on transferring the hound-python repo to the thoughtbot org? We'd like to have some control over it since we'll be running the service and maintaining it. As mentioned previously, we'll happily add you as a collaborator.

jmcarp commented:
@salbertson @thorncp sorry, didn't see this. Sure, I'd be happy to transfer the repo. But it looks like I can't because I don't have admin rights to the organization. Does GitHub offer a way around this?

teoljungberg commented:
@jmcarp If you add either one of us on @thoughtbot/hound (@salbertson, @thorncp, @gylaz, or myself) - we can transfer it to the thoughtbot organization

gylaz commented:
@jakecraige Any progress on this? We'd like to try and get this all hooked up.

croaky commented:
Does the repo need to be transferred or would it suffice to clone and push to the new org? Benefit being we'd remove a blocker to getting this out.

gylaz commented:
> Does the repo need to be transferred or would it suffice to clone and push to the new org? Benefit being we'd remove a blocker to getting this out.

I assume you mean the proposed `houndci` org. This is certainly an alternative (or forking), if @jmcarp is okay with that rather than transferring.

thorncp commented:
I think the current blocker for us is that the repo doesn't have a license attached to it, so it defaults to @jmcarp having full copyrights and we can't fork it.

croaky commented:
If the blocker is legal stuff, can we provide some suggested text for the LICENSE file (or other docs) as a GitHub comment?

jmcarp commented:
Added @teoljungberg as a contributor to hound-python.

thorncp commented:
Awesome, thanks @jmcarp !

mike-burns commented:
@jmcarp https://github.com/thoughtbot/hound-python/blob/master/review.py#L9 - can you explain this TODO a bit more, especially so that we know when we should take action?
